failed to write bulk sync blocks: failed to verify the header: not enough seals to seal block

## failed to write bulk sync blocks: failed to verify the header: not enough seals to seal block

### Description
- We pulled in latest develop changes to one of our validator and are seeing this error.



### Your environment

* OS and version: `Ubuntu 20`
* version of the Polygon Edge: `dd1d9af9d602fe1c88ce473c09259a038d07a16d`
* branch that causes this issue: `develop`

### Steps to reproduce

- Create a validator node with latest develop changes


### Logs

```
May 11 14:57:23 ip-172-31-4-5 main[579319]: 2022-05-11T14:57:23.511Z [ERROR] polygon.consensus.ibft: failed to bulk sync: err="failed to write bulk sync blocks: failed to verify the header: not enough seals to seal block"
```

we suspect [https://github.com/0xPolygon/polygon-edge/pull/513](https://github.com/0xPolygon/polygon-edge/pull/513) this PR might have introduced this error

> Hey @akshar , thank you for raising the issue.
> 
> #513 is what introduced the issue most likely. The returned value from `QuorumSize()` is different than before for the same size of the validator set. In effect, this means that previously sealed blocks (verified under the old calculation) cannot be verified with the new calculation during bulk sync.
> 
> 1. Do you have a public JSON-RPC endpoint ?
- `https://rpc.toronto.sx.technology`
> 2. How many validators are in the network ?
- `5`

```
[VALIDATORS]
ADDRESS
0xA3a585c6d59CCE6aAe7035e8df48b3327cC8BE54
0x7EFF16a34e3433182D636488bc97919b10283F37
0x3e64F88C6C7a1310236B242180c0Ba1409d10F4d
0x9cde4F11EA57163023A9a8a5dF499b27813639EB
0x6c230a5f8dd3413aA04e0860dD225F887b36f101
```


> 3. What is block number that cannot be verified ?
```
May 12 21:17:11 ip-172-31-11-88 main[312344]: 2022-05-12T21:17:11.307Z [DEBUG] polygon.consensus.syncer: fork found: ancestor=4014669
May 12 21:17:11 ip-172-31-11-88 main[312344]: 2022-05-12T21:17:11.307Z [DEBUG] polygon.consensus.syncer: sync up to block: from=4014670 to=4016340
May 12 21:17:11 ip-172-31-11-88 main[312344]: 2022-05-12T21:17:11.313Z [INFO]  polygon.blockchain: write block: num=4014670 parent=0xb525cb4c5deffdbc069b6215819281d7d86cee0fdba0f3bc78ec3d490cfce6e5
May 12 21:17:11 ip-172-31-11-88 main[312344]: 2022-05-12T21:17:11.314Z [ERROR] polygon.consensus.ibft: failed to bulk sync: err="failed to write bulk sync blocks: failed to verify the header: not enough seals to seal block"
May 12 21:17:11 ip-172-31-11-88 main[312344]: 2022-05-12T21:17:11.345Z [DEBUG] polygon.consensus.syncer: fork found: ancestor=4014669
May 12 21:17:11 ip-172-31-11-88 main[312344]: 2022-05-12T21:17:11.345Z [DEBUG] polygon.consensus.syncer: sync up to block: from=4014670 to=4016340
May 12 21:17:11 ip-172-31-11-88 main[312344]: 2022-05-12T21:17:11.355Z [INFO]  polygon.blockchain: write block: num=4014670 parent=0xb525cb4c5deffdbc069b6215819281d7d86cee0fdba0f3bc78ec3d490cfce6e5
May 12 21:17:11 ip-172-31-11-88 main[312344]: 2022-05-12T21:17:11.357Z [ERROR] polygon.consensus.ibft: failed to bulk sync: err="failed to write bulk sync blocks: failed to verify the header: not enough seals to seal block"
May 12 21:17:11 ip-172-31-11-88 main[312344]: 2022-05-12T21:17:11.380Z [DEBUG] polygon.consensus.syncer: fork found: ancestor=4014669

May 12 21:17:11 ip-172-31-11-88 main[312344]: 2022-05-12T21:17:11.380Z [DEBUG] polygon.consensus.syncer: sync up to block: from=4014670 to=4016340
May 12 21:17:11 ip-172-31-11-88 main[312344]: 2022-05-12T21:17:11.390Z [INFO]  polygon.blockchain: write block: num=4014670 parent=0xb525cb4c5deffdbc069b6215819281d7d86cee0fdba0f3bc78ec3d490cfce6e5

May 12 21:17:11 ip-172-31-11-88 main[312344]: 2022-05-12T21:17:11.392Z [ERROR] polygon.consensus.ibft: failed to bulk sync: err="failed to write bulk sync blocks: failed to verify the header: not enough seals to seal block"
May 12 21:17:11 ip-172-31-11-88 main[312344]: 2022-05-12T21:17:11.414Z [DEBUG] polygon.consensus.syncer: fork found: ancestor=4014669

May 12 21:17:11 ip-172-31-11-88 main[312344]: 2022-05-12T21:17:11.414Z [DEBUG] polygon.consensus.syncer: sync up to block: from=4014670 to=4016340

May 12 21:17:11 ip-172-31-11-88 main[312344]: 2022-05-12T21:17:11.424Z [INFO]  polygon.blockchain: write block: num=4014670 parent=0xb525cb4c5deffdbc069b6215819281d7d86cee0fdba0f3bc78ec3d490cfce6e5
May 12 21:17:11 ip-172-31-11-88 main[312344]: 2022-05-12T21:17:11.426Z [ERROR] polygon.consensus.ibft: failed to bulk sync: err="failed to write bulk sync blocks: failed to verify the header: not enough seals to seal block"
```
- we were consistently getting this error on all validator nodes (all of them had develop changes)


> 4. If possible, provide the output of `eth_getBlockByNumber` so I can see who are the validators that committed the seals

```
{
    "jsonrpc": "2.0",
    "id": 1,
    "result": {
        "parentHash": "0xb525cb4c5deffdbc069b6215819281d7d86cee0fdba0f3bc78ec3d490cfce6e5",
        "sha3Uncles": "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
        "miner": "0x0000000000000000000000000000000000000000",
        "stateRoot": "0x04812af12359f85e0ac0b7209d951471ccd7e932f6cb346150b7b31b4440e8bc",
        "transactionsRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
        "receiptsRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
        "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
        "difficulty": "0x3d424e",
        "totalDifficulty": "0x3d424e",
        "size": "0x3a1",
        "number": "0x3d424e",
        "gasLimit": "0x3938700",
        "gasUsed": "0x0",
        "timestamp": "0x627d3c29",
        "extraData": "0x0000000000000000000000000000000000000000000000000000000000000000f90179f86994a3a585c6d59cce6aae7035e8df48b3327cc8be54947eff16a34e3433182d636488bc97919b10283f37943e64f88c6c7a1310236b242180c0ba1409d10f4d949cde4f11ea57163023a9a8a5df499b27813639eb946c230a5f8dd3413aa04e0860dd225f887b36f101b84111f5423531cf6e67c029b4953f6aaaca6e5aa918b315ec582f1704b7ce2b732b5a810af0bcb3f04495a8d2c69403aa7e777722c22a9c2e6838fa87e57faab08201f8c9b841c0e28bb611ee16cba559f10c821eaeb22e135dc58ca13d2e7fcbe8c886721058489fc55a5c9b8fac51fa5b6177d2a979aa4b0e3c582e907a80471bdab6cac86301b8417269fdfda39d5b267701d1c032ea69d2d4fe34671f0ad0ae7e04079ea942249955367649b40127bfe47f732452edae2de43c017199e1b1a35355a3216f240ea401b8417fa48ff637af1c085da5394a2d7c664293e213373d0676a0c1c474ec949583ce0d0b5703e4520b506e9f163bea8808fd99185d9540e43b3420bd220dfd430ccd01",
        "mixHash": "0x63746963616c2062797a616e74696e65206661756c7420746f6c6572616e6365",
        "nonce": "0x0000000000000000",
        "hash": "0x4ab55a70e523e514ea5cd5e97a57ed46f6fda62aa3991062738b541a9fefd82a",
        "transactions": [],
        "uncles": []
    }
}
```

- we see a new [PR](https://github.com/0xPolygon/polygon-edge/pull/549) related to quorum. Is it related to this issue ? If yes, can we get some context ? @dbrajovic

--- a/command/ibft/ibft.go
+++ b/command/ibft/ibft.go
@@ -36,7 +36,7 @@ func registerSubcommands(baseCmd *cobra.Command) {
 		candidates.GetCommand(),
 		// ibft switch
 		_switch.GetCommand(),
-		//	ibft quorum
+		// ibft quorum
 		quorum.GetCommand(),
 	)
 }

--- a/command/ibft/ibft.go
+++ b/command/ibft/ibft.go
@@ -4,6 +4,7 @@ import (
 	"github.com/0xPolygon/polygon-edge/command/helper"
 	"github.com/0xPolygon/polygon-edge/command/ibft/candidates"
 	"github.com/0xPolygon/polygon-edge/command/ibft/propose"
+	"github.com/0xPolygon/polygon-edge/command/ibft/quorum"
 	"github.com/0xPolygon/polygon-edge/command/ibft/snapshot"
 	"github.com/0xPolygon/polygon-edge/command/ibft/status"
 	_switch "github.com/0xPolygon/polygon-edge/command/ibft/switch"
@@ -35,5 +36,7 @@ func registerSubcommands(baseCmd *cobra.Command) {
 		candidates.GetCommand(),
 		// ibft switch
 		_switch.GetCommand(),
+		//	ibft quorum
+		quorum.GetCommand(),
 	)
 }

--- a/command/ibft/quorum/ibft_quorum.go
+++ b/command/ibft/quorum/ibft_quorum.go
@@ -25,9 +25,7 @@ func setFlags(cmd *cobra.Command) {
 		&params.genesisPath,
 		chainFlag,
 		fmt.Sprintf("./%s", command.DefaultGenesisFileName),
-		fmt.Sprintf(
-			"the genesis file to update",
-		),
+		"the genesis file to update",
 	)
 
 	cmd.Flags().Uint64Var(

--- a/command/ibft/quorum/ibft_quorum.go
+++ b/command/ibft/quorum/ibft_quorum.go
@@ -26,7 +26,7 @@ func setFlags(cmd *cobra.Command) {
 		chainFlag,
 		fmt.Sprintf("./%s", command.DefaultGenesisFileName),
 		fmt.Sprintf(
-			"the genesis file to update. Default: ./%s",
+			"the genesis file to update: ./%s",
 			command.DefaultGenesisFileName,
 		),
 	)

--- a/command/ibft/quorum/ibft_quorum.go
+++ b/command/ibft/quorum/ibft_quorum.go
@@ -26,8 +26,7 @@ func setFlags(cmd *cobra.Command) {
 		chainFlag,
 		fmt.Sprintf("./%s", command.DefaultGenesisFileName),
 		fmt.Sprintf(
-			"the genesis file to update: ./%s",
-			command.DefaultGenesisFileName,
+			"the genesis file to update",
 		),
 	)
 

--- a/command/ibft/quorum/ibft_quorum.go
+++ b/command/ibft/quorum/ibft_quorum.go
@@ -1,0 +1,69 @@
+package quorum
+
+import (
+	"fmt"
+	"github.com/0xPolygon/polygon-edge/command"
+	"github.com/spf13/cobra"
+)
+
+func GetCommand() *cobra.Command {
+	ibftQuorumCmd := &cobra.Command{
+		Use:     "quorum",
+		Short:   "Specify the block number after which quorum optimal will be used for reaching consensus",
+		PreRunE: runPreRun,
+		Run:     runCommand,
+	}
+
+	setFlags(ibftQuorumCmd)
+	setRequiredFlags(ibftQuorumCmd)
+
+	return ibftQuorumCmd
+}
+
+func setFlags(cmd *cobra.Command) {
+	cmd.Flags().StringVar(
+		&params.genesisPath,
+		chainFlag,
+		fmt.Sprintf("./%s", command.DefaultGenesisFileName),
+		fmt.Sprintf(
+			"the genesis file to update. Default: ./%s",
+			command.DefaultGenesisFileName,
+		),
+	)
+
+	cmd.Flags().Uint64Var(
+		&params.from,
+		fromFlag,
+		0,
+		"the height to switch the quorum calculation",
+	)
+}
+
+func setRequiredFlags(cmd *cobra.Command) {
+	for _, requiredFlag := range params.getRequiredFlags() {
+		_ = cmd.MarkFlagRequired(requiredFlag)
+	}
+}
+
+func runPreRun(_ *cobra.Command, _ []string) error {
+	return params.initRawParams()
+}
+
+func runCommand(cmd *cobra.Command, _ []string) {
+	outputter := command.InitializeOutputter(cmd)
+	defer outputter.WriteOutput()
+
+	if err := params.updateGenesisConfig(); err != nil {
+		outputter.SetError(err)
+
+		return
+	}
+
+	if err := params.overrideGenesisConfig(); err != nil {
+		outputter.SetError(err)
+
+		return
+	}
+
+	outputter.SetCommandResult(params.getResult())
+}

--- a/command/ibft/quorum/params.go
+++ b/command/ibft/quorum/params.go
@@ -41,11 +41,7 @@ func (p *quorumParams) initChain() error {
 }
 
 func (p *quorumParams) initRawParams() error {
-	if err := p.initChain(); err != nil {
-		return err
-	}
-
-	return nil
+	return p.initChain()
 }
 
 func (p *quorumParams) getRequiredFlags() []string {

--- a/command/ibft/quorum/params.go
+++ b/command/ibft/quorum/params.go
@@ -1,0 +1,104 @@
+package quorum
+
+import (
+	"errors"
+	"fmt"
+	"github.com/0xPolygon/polygon-edge/chain"
+	"github.com/0xPolygon/polygon-edge/command"
+	"github.com/0xPolygon/polygon-edge/command/helper"
+	"github.com/0xPolygon/polygon-edge/helper/common"
+	"os"
+)
+
+const (
+	fromFlag  = "from"
+	chainFlag = "chain"
+)
+
+var (
+	params = &quorumParams{}
+)
+
+type quorumParams struct {
+	genesisConfig *chain.Chain
+	from          uint64
+	genesisPath   string
+}
+
+func (p *quorumParams) initChain() error {
+	cc, err := chain.Import(p.genesisPath)
+	if err != nil {
+		return fmt.Errorf(
+			"failed to load chain config from %s: %w",
+			p.genesisPath,
+			err,
+		)
+	}
+
+	p.genesisConfig = cc
+
+	return nil
+}
+
+func (p *quorumParams) initRawParams() error {
+	if err := p.initChain(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *quorumParams) getRequiredFlags() []string {
+	return []string{
+		fromFlag,
+	}
+}
+
+func (p *quorumParams) updateGenesisConfig() error {
+	return appendIBFTQuorum(
+		p.genesisConfig,
+		p.from,
+	)
+}
+
+func (p *quorumParams) overrideGenesisConfig() error {
+	// Remove the current genesis configuration from disk
+	if err := os.Remove(p.genesisPath); err != nil {
+		return err
+	}
+
+	// Save the new genesis configuration
+	if err := helper.WriteGenesisConfigToDisk(
+		p.genesisConfig,
+		p.genesisPath,
+	); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *quorumParams) getResult() command.CommandResult {
+	result := &IBFTQuorumResult{
+		Chain: p.genesisPath,
+		From:  common.JSONNumber{Value: p.from},
+	}
+
+	return result
+}
+
+func appendIBFTQuorum(
+	cc *chain.Chain,
+	from uint64,
+) error {
+	ibftConfig, ok := cc.Params.Engine["ibft"].(map[string]interface{})
+	if !ok {
+		return errors.New(`"ibft" setting doesn't exist in "engine" of genesis.json'`)
+	}
+
+	ibftConfig["quorumSizeBlockNum"] = from
+
+	cc.Params.Engine["ibft"] = ibftConfig
+
+	return nil
+}

--- a/command/ibft/quorum/params.go
+++ b/command/ibft/quorum/params.go
@@ -79,12 +79,10 @@ func (p *quorumParams) overrideGenesisConfig() error {
 }
 
 func (p *quorumParams) getResult() command.CommandResult {
-	result := &IBFTQuorumResult{
+	return &IBFTQuorumResult{
 		Chain: p.genesisPath,
 		From:  common.JSONNumber{Value: p.from},
 	}
-
-	return result
 }
 
 func appendIBFTQuorum(

--- a/command/ibft/quorum/result.go
+++ b/command/ibft/quorum/result.go
@@ -1,0 +1,29 @@
+package quorum
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/0xPolygon/polygon-edge/command/helper"
+	"github.com/0xPolygon/polygon-edge/helper/common"
+)
+
+type IBFTQuorumResult struct {
+	Chain string            `json:"chain"`
+	From  common.JSONNumber `json:"from"`
+}
+
+func (r *IBFTQuorumResult) GetOutput() string {
+	var buffer bytes.Buffer
+
+	buffer.WriteString("\n[NEW IBFT QUORUM START]\n")
+
+	outputs := []string{
+		fmt.Sprintf("Chain|%s", r.Chain),
+		fmt.Sprintf("From|%d", r.From.Value),
+	}
+
+	buffer.WriteString(helper.FormatKV(outputs))
+	buffer.WriteString("\n")
+
+	return buffer.String()
+}

--- a/consensus/ibft/ibft.go
+++ b/consensus/ibft/ibft.go
@@ -1256,6 +1256,9 @@ func (i *Ibft) VerifyHeader(parent, header *types.Header) error {
 	return nil
 }
 
+//	quorumSize returns a callback that when executed on a ValidatorSet computes
+//	number of votes required to reach quorum based on the size of the set.
+//	The blockNumber argument indicates which formula was used to calculate the result (see PRs #513, #549)
 func (i *Ibft) quorumSize(blockNumber uint64) QuorumImplementation {
 	rawUint64, ok := i.config.Config["quorumSizeBlockNum"]
 	if !ok {

--- a/consensus/ibft/ibft.go
+++ b/consensus/ibft/ibft.go
@@ -1269,9 +1269,9 @@ func (i *Ibft) quorumSize(blockNumber uint64) QuorumImplementation {
 
 	if blockNumber < uint64(cfgQuorumSizeBlockNum) {
 		return LegacyQuorumSize
-	} else {
-		return OptimalQuorumSize
 	}
+
+	return OptimalQuorumSize
 }
 
 // ProcessHeaders updates the snapshot based on previously verified headers

--- a/consensus/ibft/ibft.go
+++ b/consensus/ibft/ibft.go
@@ -1256,6 +1256,24 @@ func (i *Ibft) VerifyHeader(parent, header *types.Header) error {
 	return nil
 }
 
+func (i *Ibft) quorumSize(validators ValidatorSet, blockNumber uint64) int {
+	rawUint64, ok := i.config.Config["quorumSizeBlockNum"]
+	if !ok {
+		return OptimalQuorumSize(validators)
+	}
+
+	cfgQuorumSizeBlockNum, ok := rawUint64.(uint64)
+	if !ok {
+		panic("bad")
+	}
+
+	if blockNumber < cfgQuorumSizeBlockNum {
+		return LegacyQuorumSize(validators)
+	} else {
+		return OptimalQuorumSize(validators)
+	}
+}
+
 // ProcessHeaders updates the snapshot based on previously verified headers
 func (i *Ibft) ProcessHeaders(headers []*types.Header) error {
 	return i.processHeaders(headers)

--- a/consensus/ibft/ibft.go
+++ b/consensus/ibft/ibft.go
@@ -78,8 +78,9 @@ type Ibft struct {
 
 	txpool txPoolInterface // Reference to the transaction pool
 
-	store     *snapshotStore // Snapshot store that keeps track of all snapshots
-	epochSize uint64
+	store              *snapshotStore // Snapshot store that keeps track of all snapshots
+	epochSize          uint64
+	quorumSizeBlockNum uint64
 
 	msgQueue *msgQueue     // Structure containing different message queues
 	updateCh chan struct{} // Update channel
@@ -133,11 +134,13 @@ func (i *Ibft) runHook(hookName HookType, height uint64, hookParam interface{}) 
 func Factory(
 	params *consensus.ConsensusParams,
 ) (consensus.Consensus, error) {
-	var epochSize uint64
-	if definedEpochSize, ok := params.Config.Config["epochSize"]; !ok {
-		// No epoch size defined, use the default one
-		epochSize = DefaultEpochSize
-	} else {
+	//	defaults for user set fields in genesis
+	var (
+		epochSize          = uint64(DefaultEpochSize)
+		quorumSizeBlockNum = uint64(0)
+	)
+
+	if definedEpochSize, ok := params.Config.Config["epochSize"]; ok {
 		// Epoch size is defined, use the passed in one
 		readSize, ok := definedEpochSize.(float64)
 		if !ok {
@@ -147,21 +150,32 @@ func Factory(
 		epochSize = uint64(readSize)
 	}
 
+	if rawBlockNum, ok := params.Config.Config["quorumSizeBlockNum"]; ok {
+		//	Block number specified for quorum size switch
+		readBlockNum, ok := rawBlockNum.(float64)
+		if !ok {
+			return nil, errors.New("invalid type assertion")
+		}
+
+		quorumSizeBlockNum = uint64(readBlockNum)
+	}
+
 	p := &Ibft{
-		logger:         params.Logger.Named("ibft"),
-		config:         params.Config,
-		Grpc:           params.Grpc,
-		blockchain:     params.Blockchain,
-		executor:       params.Executor,
-		closeCh:        make(chan struct{}),
-		txpool:         params.Txpool,
-		state:          &currentState{},
-		network:        params.Network,
-		epochSize:      epochSize,
-		sealing:        params.Seal,
-		metrics:        params.Metrics,
-		secretsManager: params.SecretsManager,
-		blockTime:      time.Duration(params.BlockTime) * time.Second,
+		logger:             params.Logger.Named("ibft"),
+		config:             params.Config,
+		Grpc:               params.Grpc,
+		blockchain:         params.Blockchain,
+		executor:           params.Executor,
+		closeCh:            make(chan struct{}),
+		txpool:             params.Txpool,
+		state:              &currentState{},
+		network:            params.Network,
+		epochSize:          epochSize,
+		quorumSizeBlockNum: quorumSizeBlockNum,
+		sealing:            params.Seal,
+		metrics:            params.Metrics,
+		secretsManager:     params.SecretsManager,
+		blockTime:          time.Duration(params.BlockTime) * time.Second,
 	}
 
 	// Initialize the mechanism
@@ -1260,17 +1274,7 @@ func (i *Ibft) VerifyHeader(parent, header *types.Header) error {
 //	number of votes required to reach quorum based on the size of the set.
 //	The blockNumber argument indicates which formula was used to calculate the result (see PRs #513, #549)
 func (i *Ibft) quorumSize(blockNumber uint64) QuorumImplementation {
-	rawUint64, ok := i.config.Config["quorumSizeBlockNum"]
-	if !ok {
-		return OptimalQuorumSize
-	}
-
-	cfgQuorumSizeBlockNum, ok := rawUint64.(float64)
-	if !ok {
-		panic("invalid block number format")
-	}
-
-	if blockNumber < uint64(cfgQuorumSizeBlockNum) {
+	if blockNumber < i.quorumSizeBlockNum {
 		return LegacyQuorumSize
 	}
 

--- a/consensus/ibft/ibft.go
+++ b/consensus/ibft/ibft.go
@@ -1262,12 +1262,12 @@ func (i *Ibft) quorumSize(blockNumber uint64) QuorumImplementation {
 		return OptimalQuorumSize
 	}
 
-	cfgQuorumSizeBlockNum, ok := rawUint64.(uint64)
+	cfgQuorumSizeBlockNum, ok := rawUint64.(float64)
 	if !ok {
-		panic("bad")
+		panic("invalid block number format")
 	}
 
-	if blockNumber < cfgQuorumSizeBlockNum {
+	if blockNumber < uint64(cfgQuorumSizeBlockNum) {
 		return LegacyQuorumSize
 	} else {
 		return OptimalQuorumSize

--- a/consensus/ibft/ibft.go
+++ b/consensus/ibft/ibft.go
@@ -916,12 +916,12 @@ func (i *Ibft) runValidateState() {
 			panic(fmt.Sprintf("BUG: %s", reflect.TypeOf(msg.Type)))
 		}
 
-		if i.state.numPrepared() >= i.state.validators.QuorumSize() {
+		if i.state.numPrepared() >= i.quorumSize(i.state.view.Sequence)(i.state.validators) {
 			// we have received enough pre-prepare messages
 			sendCommit()
 		}
 
-		if i.state.numCommitted() >= i.state.validators.QuorumSize() {
+		if i.state.numCommitted() >= i.quorumSize(i.state.view.Sequence)(i.state.validators) {
 			// we have received enough commit messages
 			sendCommit()
 
@@ -1107,7 +1107,7 @@ func (i *Ibft) runRoundChangeState() {
 			// update timer
 			timeout = exponentialTimeout(i.state.view.Round)
 			sendRoundChange(msg.View.Round)
-		} else if num == i.state.validators.QuorumSize() {
+		} else if num == i.quorumSize(i.state.view.Sequence)(i.state.validators) {
 			// start a new round immediately
 			i.state.view.Round = msg.View.Round
 			i.setState(AcceptState)
@@ -1249,7 +1249,7 @@ func (i *Ibft) VerifyHeader(parent, header *types.Header) error {
 	}
 
 	// verify the committed seals
-	if err := verifyCommitedFields(snap, header); err != nil {
+	if err := verifyCommitedFields(snap, header, i.quorumSize(header.Number)); err != nil {
 		return err
 	}
 

--- a/consensus/ibft/ibft.go
+++ b/consensus/ibft/ibft.go
@@ -1256,10 +1256,10 @@ func (i *Ibft) VerifyHeader(parent, header *types.Header) error {
 	return nil
 }
 
-func (i *Ibft) quorumSize(validators ValidatorSet, blockNumber uint64) int {
+func (i *Ibft) quorumSize(blockNumber uint64) QuorumImplementation {
 	rawUint64, ok := i.config.Config["quorumSizeBlockNum"]
 	if !ok {
-		return OptimalQuorumSize(validators)
+		return OptimalQuorumSize
 	}
 
 	cfgQuorumSizeBlockNum, ok := rawUint64.(uint64)
@@ -1268,9 +1268,9 @@ func (i *Ibft) quorumSize(validators ValidatorSet, blockNumber uint64) int {
 	}
 
 	if blockNumber < cfgQuorumSizeBlockNum {
-		return LegacyQuorumSize(validators)
+		return LegacyQuorumSize
 	} else {
-		return OptimalQuorumSize(validators)
+		return OptimalQuorumSize
 	}
 }
 

--- a/consensus/ibft/ibft_test.go
+++ b/consensus/ibft/ibft_test.go
@@ -1303,7 +1303,7 @@ func TestQuorumSizeSwitch(t *testing.T) {
 
 	testTable := []struct {
 		name           string
-		switchBlock    float64
+		switchBlock    uint64
 		currentBlock   uint64
 		set            ValidatorSet
 		expectedQuorum int
@@ -1344,11 +1344,7 @@ func TestQuorumSizeSwitch(t *testing.T) {
 			t.Parallel()
 
 			ibft := &Ibft{
-				config: &consensus.Config{
-					Config: map[string]interface{}{
-						"quorumSizeBlockNum": test.switchBlock,
-					},
-				},
+				quorumSizeBlockNum: test.switchBlock,
 			}
 
 			assert.Equal(t,

--- a/consensus/ibft/ibft_test.go
+++ b/consensus/ibft/ibft_test.go
@@ -1303,7 +1303,7 @@ func TestQuorumSizeSwitch(t *testing.T) {
 
 	testTable := []struct {
 		name           string
-		switchBlock    uint64
+		switchBlock    float64
 		currentBlock   uint64
 		set            ValidatorSet
 		expectedQuorum int

--- a/consensus/ibft/ibft_test.go
+++ b/consensus/ibft/ibft_test.go
@@ -1297,3 +1297,64 @@ func TestGetIBFTForks(t *testing.T) {
 		})
 	}
 }
+
+func TestQuorumSizeSwitch(t *testing.T) {
+	t.Parallel()
+
+	testTable := []struct {
+		name           string
+		switchBlock    uint64
+		currentBlock   uint64
+		set            ValidatorSet
+		expectedQuorum int
+	}{
+		{
+			"use old quorum calculation",
+			10,
+			5,
+			[]types.Address{
+				types.ZeroAddress,
+				types.ZeroAddress,
+				types.ZeroAddress,
+				types.ZeroAddress,
+				types.ZeroAddress,
+				types.ZeroAddress,
+			},
+			3,
+		},
+		{
+			"use new quorum calculation",
+			10,
+			15,
+			[]types.Address{
+				types.ZeroAddress,
+				types.ZeroAddress,
+				types.ZeroAddress,
+				types.ZeroAddress,
+				types.ZeroAddress,
+				types.ZeroAddress,
+			},
+			4,
+		},
+	}
+
+	for _, test := range testTable {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+
+			ibft := &Ibft{
+				config: &consensus.Config{
+					Config: map[string]interface{}{
+						"quorumSizeBlockNum": test.switchBlock,
+					},
+				},
+			}
+
+			assert.Equal(t,
+				test.expectedQuorum,
+				ibft.quorumSize(test.currentBlock)(test.set),
+			)
+		})
+	}
+}

--- a/consensus/ibft/sign.go
+++ b/consensus/ibft/sign.go
@@ -166,7 +166,11 @@ func verifySigner(snap *Snapshot, header *types.Header) error {
 }
 
 // verifyCommitedFields is checking for consensus proof in the header
-func verifyCommitedFields(snap *Snapshot, header *types.Header) error {
+func verifyCommitedFields(
+	snap *Snapshot,
+	header *types.Header,
+	quorumSizeFn QuorumImplementation,
+) error {
 	extra, err := getIbftExtra(header)
 	if err != nil {
 		return err
@@ -207,7 +211,7 @@ func verifyCommitedFields(snap *Snapshot, header *types.Header) error {
 	// Valid committed seals must be at least 2F+1
 	// 	2F 	is the required number of honest validators who provided the committed seals
 	// 	+1	is the proposer
-	if validSeals := len(visited); validSeals < snap.Set.QuorumSize() {
+	if validSeals := len(visited); validSeals < quorumSizeFn(snap.Set) {
 		return fmt.Errorf("not enough seals to seal block")
 	}
 

--- a/consensus/ibft/sign_test.go
+++ b/consensus/ibft/sign_test.go
@@ -59,7 +59,7 @@ func TestSign_CommittedSeals(t *testing.T) {
 
 		assert.NoError(t, err)
 
-		return verifyCommitedFields(snap, sealed)
+		return verifyCommitedFields(snap, sealed, OptimalQuorumSize)
 	}
 
 	// Correct

--- a/consensus/ibft/state.go
+++ b/consensus/ibft/state.go
@@ -304,19 +304,26 @@ func (v *ValidatorSet) MaxFaultyNodes() int {
 	return (len(*v) - 1) / 3
 }
 
-// QuorumSize returns the number of required messages for consensus
-func (v ValidatorSet) QuorumSize() int {
+//	LegacyQuorumSize returns the legacy quorum size for the given validator set
+func LegacyQuorumSize(set ValidatorSet) int {
+	// According to the IBFT spec, the number of valid messages
+	// needs to be 2F + 1
+	return 2*set.MaxFaultyNodes() + 1
+}
+
+// OptimalQuorumSize returns the optimal quorum size for the given validator set
+func OptimalQuorumSize(set ValidatorSet) int {
 	//	if the number of validators is less than 4,
 	//	then the entire set is required
-	if v.MaxFaultyNodes() == 0 {
+	if set.MaxFaultyNodes() == 0 {
 		/*
 			N: 1 -> Q: 1
 			N: 2 -> Q: 2
 			N: 3 -> Q: 3
 		*/
-		return v.Len()
+		return set.Len()
 	}
 
 	// (quorum optimal)	Q = ceil(2/3 * N)
-	return int(math.Ceil(2 * float64(v.Len()) / 3))
+	return int(math.Ceil(2 * float64(set.Len()) / 3))
 }

--- a/consensus/ibft/state.go
+++ b/consensus/ibft/state.go
@@ -304,6 +304,8 @@ func (v *ValidatorSet) MaxFaultyNodes() int {
 	return (len(*v) - 1) / 3
 }
 
+type QuorumImplementation func(ValidatorSet) int
+
 //	LegacyQuorumSize returns the legacy quorum size for the given validator set
 func LegacyQuorumSize(set ValidatorSet) int {
 	// According to the IBFT spec, the number of valid messages

--- a/consensus/ibft/state_test.go
+++ b/consensus/ibft/state_test.go
@@ -62,7 +62,7 @@ func TestNumValid(t *testing.T) {
 
 		assert.Equal(t,
 			int(c.Quorum),
-			pool.ValidatorSet().QuorumSize(),
+			OptimalQuorumSize(pool.ValidatorSet()),
 		)
 	}
 }

